plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkLaunchConfigurationDelegate.java

```diff
@@ -0,0 +1,233 @@
+ * the License.
+ *******************************************************************************/
+package com.google.cloud.tools.eclipse.appengine.localserver.server;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.maven.project.MavenProject;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.debug.core.DebugPlugin;
+import org.eclipse.debug.core.ILaunch;
+import org.eclipse.debug.core.ILaunchConfiguration;
+import org.eclipse.debug.core.ILaunchConfigurationType;
+import org.eclipse.debug.core.ILaunchConfigurationWorkingCopy;
+import org.eclipse.debug.core.ILaunchManager;
+import org.eclipse.jdt.launching.AbstractJavaLaunchConfigurationDelegate;
+import org.eclipse.jdt.launching.IJavaLaunchConfigurationConstants;
+import org.eclipse.jdt.launching.SocketUtil;
+import org.eclipse.m2e.core.MavenPlugin;
+import org.eclipse.m2e.core.project.IMavenProjectFacade;
+import org.eclipse.wst.server.core.IModule;
+import org.eclipse.wst.server.core.IRuntime;
+import org.eclipse.wst.server.core.IServer;
+import org.eclipse.wst.server.core.IServerListener;
+import org.eclipse.wst.server.core.ServerEvent;
+import org.eclipse.wst.server.core.ServerUtil;
+
+import com.google.cloud.tools.eclipse.appengine.localserver.Activator;
+import com.google.cloud.tools.eclipse.appengine.localserver.GCloudCommandDelegate;
+
+/**
+ * Cloud SDK server's launch configuration delegate.
+ */
+public class CloudSdkLaunchConfigurationDelegate extends AbstractJavaLaunchConfigurationDelegate {
+  private IServerListener debugServerListener;
+  private static final String DEBUGGER_HOST = "localhost";
+
+  @Override
+  public void launch(ILaunchConfiguration configuration,
+                     String mode,
+                     ILaunch launch,
+                     IProgressMonitor monitor) throws CoreException {
+    final IServer server = ServerUtil.getServer(configuration);
+    if (server == null) {
+      return;
+    }
+
+    final IModule[] modules = server.getModules();
+    if (modules == null || modules.length == 0) {
+      return;
+    }
+
+    String runnables = getRunnable(modules[0].getProject(), monitor);
+    IRuntime runtime = server.getRuntime();
+    if (runtime == null) {
+      return;
+    }
+    IPath sdkLocation = runtime.getLocation();
+
+    CloudSdkServerBehaviour sb = (CloudSdkServerBehaviour) server.loadAdapter(CloudSdkServerBehaviour.class,
+                                                                              null);
+    sb.setupLaunch(mode);
+
+    final CloudSdkServer cloudSdkServer = CloudSdkServer.getCloudSdkServer(server);
+    int debugPort = -1;
+    if (mode.equals(ILaunchManager.DEBUG_MODE)) {
```

Comment:
some devs like to put constants on the left side of .equals to avoid NullPointerExceptions